File: mm-php/src/io/metamessage/mm/MmValidator.php
<?php

namespace io\metamessage\mm;

class ValidationResult {
    public bool $isValid;
    public array $errors;
    
    public function __construct() {
        $this->isValid = true;
        $this->errors = [];
    }
    
    public function addError(string $error): void {
        $this->isValid = false;
        $this->errors[] = $error;
    }
}

class MmValidator {
    public function validate(mixed $value, MmTag $tag): ValidationResult {
        $result = new ValidationResult();

        if ($value === null) {
            if (!$tag->nullable) {
                $result->addError('value is required');
            }
            return $result;
        }

        switch ($tag->type) {
            case ValueType::BOOL:
                $this->validateBool($value, $tag, $result);
                break;
            case ValueType::INT:
                $this->validateInt($value, $tag, $result);
                break;
            case ValueType::INT8:
                $this->validateInt8($value, $tag, $result);
                break;
            case ValueType::INT16:
                $this->validateInt16($value, $tag, $result);
                break;
            case ValueType::INT32:
                $this->validateInt32($value, $tag, $result);
                break;
            case ValueType::INT64:
                $this->validateInt64($value, $tag, $result);
                break;
            case ValueType::UINT:
                $this->validateUint($value, $tag, $result);
                break;
            case ValueType::UINT8:
                $this->validateUint8($value, $tag, $result);
                break;
            case ValueType::UINT16:
                $this->validateUint16($value, $tag, $result);
                break;
            case ValueType::UINT32:
                $this->validateUint32($value, $tag, $result);
                break;
            case ValueType::UINT64:
                $this->validateUint64($value, $tag, $result);
                break;
            case ValueType::BIGINT:
                $this->validateBigInt($value, $tag, $result);
                break;
            case ValueType::FLOAT32:
                $this->validateFloat32($value, $tag, $result);
                break;
            case ValueType::FLOAT64:
                $this->validateFloat64($value, $tag, $result);
                break;
            case ValueType::DECIMAL:
                $this->validateFloat($value, $tag, $result);
                break;
            case ValueType::STRING:
                $this->validateString($value, $tag, $result);
                break;
            case ValueType::EMAIL:
                $this->validateEmail($value, $tag, $result);
                break;
            case ValueType::URL:
                $this->validateUrl($value, $tag, $result);
                break;
            case ValueType::BYTES:
                $this->validateBytes($value, $tag, $result);
                break;
            case ValueType::UUID:
                $this->validateUuid($value, $tag, $result);
                break;
            case ValueType::DATETIME:
            case ValueType::DATE:
            case ValueType::TIME:
                $this->validateDateTime($value, $tag, $result);
                break;
            case ValueType::ENUM:
                $this->validateEnum($value, $tag, $result);
                break;
            case ValueType::ARRAY:
            case ValueType::SLICE:
                $this->validateArray($value, $tag, $result);
                break;
            case ValueType::STRUCT:
                $this->validateStruct($value, $tag, $result);
                break;
            default:
                break;
        }

        return $result;
    }

    private function validateBool(mixed $value, MmTag $tag, ValidationResult $result): void {
        if (!is_bool($value)) {
            $result->addError('value must be a boolean');
        }
    }

    private function validateInt(mixed $value, MmTag $tag, ValidationResult $result): void {
        if (!is_int($value)) {
            $result->addError('value must be a number');
        }
    }

    private function validateInt8(mixed $value, MmTag $tag, ValidationResult $result): void {
        if (!is_int($value)) {
            $result->addError('value must be a number');
            return;
        }
        if ($value < -128 || $value > 127) {
            $result->addError('value out of range for int8');
        }
    }

    private function validateInt16(mixed $value, MmTag $tag, ValidationResult $result): void {
        if (!is_int($value)) {
            $result->addError('value must be a number');
            return;
        }
        if ($value < -32768 || $value > 32767) {
            $result->addError('value out of range for int16');
        }
    }

    private function validateInt32(mixed $value, MmTag $tag, ValidationResult $result): void {
        if (!is_int($value)) {
            $result->addError('value must be a number');
            return;
        }
        if ($value < -2147483648 || $value > 2147483647) {
            $result->addError('value out of range for int32');
        }
    }

    private function validateInt64(mixed $value, MmTag $tag, ValidationResult $result): void {
        if (!is_int($value)) {
            $result->addError('value must be a number');
            return;
        }
    }

    private function validateUint(mixed $value, MmTag $tag, ValidationResult $result): void {
        if (!is_int($value)) {
            $result->addError('value must be a number');
            return;
        }
        if ($value < 0 || $value > 4294967295) {
            $result->addError('value out of range for uint');
        }
    }

    private function validateUint8(mixed $value, MmTag $tag, ValidationResult $result): void {
        if (!is_int($value)) {
            $result->addError('value must be a number');
            return;
        }
        if ($value < 0 || $value > 255) {
            $result->addError('value out of range for uint8');
        }
    }

    private function validateUint16(mixed $value, MmTag $tag, ValidationResult $result): void {
        if (!is_int($value)) {
            $result->addError('value must be a number');
            return;
        }
        if ($value < 0 || $value > 65535) {
            $result->addError('value out of range for uint16');
        }
    }

    private function validateUint32(mixed $value, MmTag $tag, ValidationResult $result): void {
        if (!is_int($value)) {
            $result->addError('value must be a number');
            return;
        }
        if ($value < 0 || $value > 4294967295) {
            $result->addError('value out of range for uint32');
        }
    }

    private function validateUint64(mixed $value, MmTag $tag, ValidationResult $result): void {
        if (!is_int($value)) {
            $result->addError('value must be a number');
            return;
        }
    }

    private function validateBigInt(mixed $value, MmTag $tag, ValidationResult $result): void {
        if (is_int($value)) {
            return;
        }
        if (!is_string($value)) {
            $result->addError('value must be a bigint string');
            return;
        }
        if (!extension_loaded('gmp')) {
            $result->addError('GMP extension is not loaded');
            return;
        }
        try {
            gmp_init($value);
        } catch (\Exception $e) {
            $result->addError('value must be a valid bigint string');
        }
    }

    private function validateFloat32(mixed $value, MmTag $tag, ValidationResult $result): void {
        if (!is_float($value) && !is_int($value)) {
            $result->addError('value must be a float');
            return;
        }
        $floatValue = (float)$value;
        if ($floatValue < -3.4028234663852886e+38 || $floatValue > 3.4028234663852886e+38) {
            $result->addError('value out of range for float32');
        }
    }

    private function validateFloat64(mixed $value, MmTag $tag, ValidationResult $result): void {
        if (!is_float($value) && !is_int($value)) {
            $result->addError('value must be a float');
            return;
        }
    }

    private function validateFloat(mixed $value, MmTag $tag, ValidationResult $result): void {
        if (!is_float($value) && !is_int($value)) {
            $result->addError('value must be a float');
        }
    }
    
    private function validateString(mixed $value, MmTag $tag, ValidationResult $result): void {
        if (!is_string($value)) {
            $result->addError('value must be a string');
            return;
        }
        
        if (empty($value) && !$tag->allowEmpty) {
            $result->addError('value is empty');
        }
    }
    
    private function validateEmail(mixed $value, MmTag $tag, ValidationResult $result): void {
        if (!is_string($value)) {
            $result->addError('value must be a string');
            return;
        }
        
        if (empty($value) && !$tag->allowEmpty) {
            $result->addError('value is empty');
            return;
        }
        
        if (!empty($value) && !filter_var($value, FILTER_VALIDATE_EMAIL)) {
            $result->addError('value is not a valid email');
        }
    }
    
    private function validateUrl(mixed $value, MmTag $tag, ValidationResult $result): void {
        if (!is_string($value)) {
            $result->addError('value must be a string');
            return;
        }
        
        if (empty($value) && !$tag->allowEmpty) {
            $result->addError('value is empty');
            return;
        }
        
        if (!empty($value) && !filter_var($value, FILTER_VALIDATE_URL)) {
            $result->addError('value is not a valid url');
        }
    }
    
    private function validateBytes(mixed $value, MmTag $tag, ValidationResult $result): void {
        if (!is_string($value) && !is_array($value)) {
            $result->addError('value must be a byte array');
            return;
        }
        
        if (empty($value) && !$tag->allowEmpty) {
            $result->addError('value is empty');
        }
    }
    
    private function validateUuid(mixed $value, MmTag $tag, ValidationResult $result): void {
        if (!is_string($value)) {
            $result->addError('value must be a string');
            return;
        }
        
        if (empty($value) && !$tag->allowEmpty) {
            $result->addError('value is empty');
            return;
        }
        
        if (!empty($value) && !preg_match('/^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i', $value)) {
            $result->addError('value is not a valid uuid');
        }
    }
    
    private function validateDateTime(mixed $value, MmTag $tag, ValidationResult $result): void {
        if (!is_string($value) && !$value instanceof \DateTime) {
            $result->addError('value must be a datetime');
        }
    }
    
    private function validateEnum(mixed $value, MmTag $tag, ValidationResult $result): void {
        if (!is_string($value)) {
            $result->addError('value must be a string');
            return;
        }
        
        if (empty($value) && !$tag->allowEmpty) {
            $result->addError('value is empty');
            return;
        }
        
        if (!empty($value) && !empty($tag->enumValues)) {
            $enumValues = explode('|', $tag->enumValues);
            $enumValues = array_map('trim', $enumValues);
            if (!in_array($value, $enumValues)) {
                $result->addError('value is not in enum');
            }
        }
    }
    
    private function validateArray(mixed $value, MmTag $tag, ValidationResult $result): void {
        if (!is_array($value)) {
            $result->addError('value must be an array');
            return;
        }
        
        if (empty($value) && !$tag->allowEmpty) {
            $result->addError('value is empty');
        }
        
        // 验证数组元素
        foreach ($value as $index => $item) {
            $childTag = MmTag::empty();
            $childTag->inheritFromArrayParent($tag);
            $itemResult = $this->validate($item, $childTag);
            if (!$itemResult->isValid) {
                foreach ($itemResult->errors as $error) {
                    $result->addError("item $index: $error");
                }
            }
        }
    }
    
    private function validateStruct(mixed $value, MmTag $tag, ValidationResult $result): void {
        if (!is_array($value) && !is_object($value)) {
            $result->addError('value must be an object');
        }
    }
}

// 全局验证器实例
$validator = new MmValidator();

File: mm-swift/Sources/MetaMessage/MM/MmValidator.swift
import Foundation

public class ValidationResult {
    public var isValid: Bool
    public var errors: [String]
    
    public init() {
        self.isValid = true
        self.errors = []
    }
    
    public func addError(_ error: String) {
        self.isValid = false
        self.errors.append(error)
    }
}

public class MmValidator {
    public init() {}
    
    public func validate(_ value: Any?, tag: JSONCTag) -> ValidationResult {
        let result = ValidationResult()
        
        if let value = value {
            switch tag.type {
            case .bool:
                validateBool(value, tag: tag, result: result)
            case .int:
                validateInt(value, tag: tag, result: result)
            case .int8:
                validateInt8(value, tag: tag, result: result)
            case .int16:
                validateInt16(value, tag: tag, result: result)
            case .int32:
                validateInt32(value, tag: tag, result: result)
            case .int64:
                validateInt64(value, tag: tag, result: result)
            case .uint:
                validateUint(value, tag: tag, result: result)
            case .uint8:
                validateUint8(value, tag: tag, result: result)
            case .uint16:
                validateUint16(value, tag: tag, result: result)
            case .uint32:
                validateUint32(value, tag: tag, result: result)
            case .uint64:
                validateUint64(value, tag: tag, result: result)
            case .bigInt:
                validateBigInt(value, tag: tag, result: result)
            case .float32:
                validateFloat32(value, tag: tag, result: result)
            case .float64, .decimal:
                validateFloat64(value, tag: tag, result: result)
            case .string:
                validateString(value, tag: tag, result: result)
            case .email:
                validateEmail(value, tag: tag, result: result)
            case .url:
                validateUrl(value, tag: tag, result: result)
            case .bytes:
                validateBytes(value, tag: tag, result: result)
            case .uuid:
                validateUUID(value, tag: tag, result: result)
            case .dateTime, .date, .time:
                validateDateTime(value, tag: tag, result: result)
            case .enumValue:
                validateEnum(value, tag: tag, result: result)
            case .array, .slice:
                validateArray(value, tag: tag, result: result)
            case .structType:
                validateStruct(value, tag: tag, result: result)
            default:
                break
            }
        } else {
            if !tag.nullable {
                result.addError("value is required")
            }
        }
        
        return result
    }
    
    private func validateBool(_ value: Any, tag: JSONCTag, result: ValidationResult) {
        if !(value is Bool) {
            result.addError("value must be a boolean")
        }
    }

    private func validateInt(_ value: Any, tag: JSONCTag, result: ValidationResult) {
        guard let intValue = value as? Int else {
            result.addError("value must be a number")
            return
        }
        if intValue == 0 && !tag.allowEmpty {
            result.addError("value is required")
            return
        }
        if intValue < Int(Int.min) || intValue > Int.max {
            result.addError("value out of range for int")
        }
        
        if let minStr = tag.min {
            if let mini = Int(minStr) {
                if intValue < mini {
                    result.addError("value \(intValue) is less than the minimum limit \(mini)")
                }
            } else {
                result.addError("failed to parse tag.min as int: \(minStr)")
            }
        }
        
        if let maxStr = tag.max {
            if let maxi = Int(maxStr) {
                if intValue > maxi {
                    result.addError("value \(intValue) exceeds the maximum limit \(maxi)")
                }
            } else {
                result.addError("failed to parse tag.max as int: \(maxStr)")
            }
        }
    }

    private func validateInt8(_ value: Any, tag: JSONCTag, result: ValidationResult) {
        guard let intValue = value as? Int else {
            result.addError("value must be a number")
            return
        }
        if intValue == 0 && !tag.allowEmpty {
            result.addError("value is required")
            return
        }
        if intValue < -128 || intValue > 127 {
            result.addError("value out of range for int8")
        }
        
        if let minStr = tag.min {
            if let mini = Int(minStr) {
                if mini < -128 || mini > 127 {
                    result.addError("tag.min \(mini) is out of int8 range [-128, 127]")
                } else if intValue < mini {
                    result.addError("value \(intValue) is less than the minimum limit \(mini)")
                }
            } else {
                result.addError("failed to parse tag.min as int8: \(minStr)")
            }
        }
        
        if let maxStr = tag.max {
            if let maxi = Int(maxStr) {
                if maxi < -128 || maxi > 127 {
                    result.addError("tag.max \(maxi) is out of int8 range [-128, 127]")
                } else if intValue > maxi {
                    result.addError("value \(intValue) exceeds the maximum limit \(maxi)")
                }
            } else {
                result.addError("failed to parse tag.max as int8: \(maxStr)")
            }
        }
    }

    private func validateInt16(_ value: Any, tag: JSONCTag, result: ValidationResult) {
        guard let intValue = value as? Int else {
            result.addError("value must be a number")
            return
        }
        if intValue == 0 && !tag.allowEmpty {
            result.addError("value is required")
            return
        }
        if intValue < -32768 || intValue > 32767 {
            result.addError("value out of range for int16")
        }
        
        if let minStr = tag.min {
            if let mini = Int(minStr) {
                if mini < -32768 || mini > 32767 {
                    result.addError("tag.min \(mini) is out of int16 range [-32768, 32767]")
                } else if intValue < mini {
                    result.addError("value \(intValue) is less than the minimum limit \(mini)")
                }
            } else {
                result.addError("failed to parse tag.min as int16: \(minStr)")
            }
        }
        
        if let maxStr = tag.max {
            if let maxi = Int(maxStr) {
                if maxi < -32768 || maxi > 32767 {
                    result.addError("tag.max \(maxi) is out of int16 range [-32768, 32767]")
                } else if intValue > maxi {
                    result.addError("value \(intValue) exceeds the maximum limit \(maxi)")
                }
            } else {
                result.addError("failed to parse tag.max as int16: \(maxStr)")
            }
        }
    }

    private func validateInt32(_ value: Any, tag: JSONCTag, result: ValidationResult) {
        guard let intValue = value as? Int else {
            result.addError("value must be a number")
            return
        }
        if intValue == 0 && !tag.allowEmpty {
            result.addError("value is required")
            return
        }
        if intValue < -2147483648 || intValue > 2147483647 {
            result.addError("value out of range for int32")
        }
        
        if let minStr = tag.min {
            if let mini = Int(minStr) {
                if mini < -2147483648 || mini > 2147483647 {
                    result.addError("tag.min \(mini) is out of int32 range [-2147483648, 2147483647]")
                } else if intValue < mini {
                    result.addError("value \(intValue) is less than the minimum limit \(mini)")
                }
            } else {
                result.addError("failed to parse tag.min as int32: \(minStr)")
            }
        }
        
        if let maxStr = tag.max {
            if let maxi = Int(maxStr) {
                if maxi < -2147483648 || maxi > 2147483647 {
                    result.addError("tag.max \(maxi) is out of int32 range [-2147483648, 2147483647]")
                } else if intValue > maxi {
                    result.addError("value \(intValue) exceeds the maximum limit \(maxi)")
                }
            } else {
                result.addError("failed to parse tag.max as int32: \(maxStr)")
            }
        }
    }

    private func validateInt64(_ value: Any, tag: JSONCTag, result: ValidationResult) {
        guard let intValue = value as? Int else {
            result.addError("value must be a number")
            return
        }
        if intValue == 0 && !tag.allowEmpty {
            result.addError("value is required")
            return
        }
        if intValue < -9223372036854775808 || intValue > 9223372036854775807 {
            result.addError("value out of range for int64")
        }
        
        if let minStr = tag.min {
            if let mini = Int(minStr) {
                if mini < -9223372036854775808 || mini > 9223372036854775807 {
                    result.addError("tag.min \(mini) is out of int64 range [-9223372036854775808, 9223372036854775807]")
                } else if intValue < mini {
                    result.addError("value \(intValue) is less than the minimum limit \(mini)")
                }
            } else {
                result.addError("failed to parse tag.min as int64: \(minStr)")
            }
        }
        
        if let maxStr = tag.max {
            if let maxi = Int(maxStr) {
                if maxi < -9223372036854775808 || maxi > 9223372036854775807 {
                    result.addError("tag.max \(maxi) is out of int64 range [-9223372036854775808, 9223372036854775807]")
                } else if intValue > maxi {
                    result.addError("value \(intValue) exceeds the maximum limit \(maxi)")
                }
            } else {
                result.addError("failed to parse tag.max as int64: \(maxStr)")
            }
        }
    }

    private func validateUint(_ value: Any, tag: JSONCTag, result: ValidationResult) {
        guard let intValue = value as? Int else {
            result.addError("value must be a number")
            return
        }
        if intValue == 0 && !tag.allowEmpty {
            result.addError("value is required")
            return
        }
        if intValue < 0 || intValue > 4294967295 {
            result.addError("value out of range for uint")
        }
        
        if let minStr = tag.min {
            if let mini = Int(minStr) {
                if mini < 0 || mini > 4294967295 {
                    result.addError("tag.min \(mini) is out of uint range [0, 4294967295]")
                } else if intValue < mini {
                    result.addError("value \(intValue) is less than the minimum limit \(mini)")
                }
            } else {
                result.addError("failed to parse tag.min as uint: \(minStr)")
            }
        }
        
        if let maxStr = tag.max {
            if let maxi = Int(maxStr) {
                if maxi < 0 || maxi > 4294967295 {
                    result.addError("tag.max \(maxi) is out of uint range [0, 4294967295]")
                } else if intValue > maxi {
                    result.addError("value \(intValue) exceeds the maximum limit \(maxi)")
                }
            } else {
                result.addError("failed to parse tag.max as uint: \(maxStr)")
            }
        }
    }

    private func validateUint8(_ value: Any, tag: JSONCTag, result: ValidationResult) {
        guard let intValue = value as? Int else {
            result.addError("value must be a number")
            return
        }
        if intValue == 0 && !tag.allowEmpty {
            result.addError("value is required")
            return
        }
        if intValue < 0 || intValue > 255 {
            result.addError("value out of range for uint8")
        }
        
        if let minStr = tag.min {
            if let mini = Int(minStr) {
                if mini < 0 || mini > 255 {
                    result.addError("tag.min \(mini) is out of uint8 range [0, 255]")
                } else if intValue < mini {
                    result.addError("value \(intValue) is less than the minimum limit \(mini)")
                }
            } else {
                result.addError("failed to parse tag.min as uint8: \(minStr)")
            }
        }
        
        if let maxStr = tag.max {
            if let maxi = Int(maxStr) {
                if maxi < 0 || maxi > 255 {
                    result.addError("tag.max \(maxi) is out of uint8 range [0, 255]")
                } else if intValue > maxi {
                    result.addError("value \(intValue) exceeds the maximum limit \(maxi)")
                }
            } else {
                result.addError("failed to parse tag.max as uint8: \(maxStr)")
            }
        }
    }

    private func validateUint16(_ value: Any, tag: JSONCTag, result: ValidationResult) {
        guard let intValue = value as? Int else {
            result.addError("value must be a number")
            return
        }
        if intValue == 0 && !tag.allowEmpty {
            result.addError("value is required")
            return
        }
        if intValue < 0 || intValue > 65535 {
            result.addError("value out of range for uint16")
        }
        
        if let minStr = tag.min {
            if let mini = Int(minStr) {
                if mini < 0 || mini > 65535 {
                    result.addError("tag.min \(mini) is out of uint16 range [0, 65535]")
                } else if intValue < mini {
                    result.addError("value \(intValue) is less than the minimum limit \(mini)")
                }
            } else {
                result.addError("failed to parse tag.min as uint16: \(minStr)")
            }
        }
        
        if let maxStr = tag.max {
            if let maxi = Int(maxStr) {
                if maxi < 0 || maxi > 65535 {
                    result.addError("tag.max \(maxi) is out of uint16 range [0, 65535]")
                } else if intValue > maxi {
                    result.addError("value \(intValue) exceeds the maximum limit \(maxi)")
                }
            } else {
                result.addError("failed to parse tag.max as uint16: \(maxStr)")
            }
        }
    }

    private func validateUint32(_ value: Any, tag: JSONCTag, result: ValidationResult) {
        guard let intValue = value as? Int else {
            result.addError("value must be a number")
            return
        }
        if intValue == 0 && !tag.allowEmpty {
            result.addError("value is required")
            return
        }
        if intValue < 0 || intValue > 4294967295 {
            result.addError("value out of range for uint32")
        }
        
        if let minStr = tag.min {
            if let mini = Int(minStr) {
                if mini < 0 || mini > 4294967295 {
                    result.addError("tag.min \(mini) is out of uint32 range [0, 4294967295]")
                } else if intValue < mini {
                    result.addError("value \(intValue) is less than the minimum limit \(mini)")
                }
            } else {
                result.addError("failed to parse tag.min as uint32: \(minStr)")
            }
        }
        
        if let maxStr = tag.max {
            if let maxi = Int(maxStr) {
                if maxi < 0 || maxi > 4294967295 {
                    result.addError("tag.max \(maxi) is out of uint32 range [0, 4294967295]")
                } else if intValue > maxi {
                    result.addError("value \(intValue) exceeds the maximum limit \(maxi)")
                }
            } else {
                result.addError("failed to parse tag.max as uint32: \(maxStr)")
            }
        }
    }

    private func validateUint64(_ value: Any, tag: JSONCTag, result: ValidationResult) {
        guard let intValue = value as? Int else {
            result.addError("value must be a number")
            return
        }
        if intValue == 0 && !tag.allowEmpty {
            result.addError("value is required")
            return
        }
        if intValue < 0 {
            result.addError("value out of range for uint64")
        }
        
        if let minStr = tag.min {
            if let mini = Int(minStr) {
                if mini < 0 {
                    result.addError("tag.min \(mini) is out of uint64 range [0, 18446744073709551615]")
                } else if intValue < mini {
                    result.addError("value \(intValue) is less than the minimum limit \(mini)")
                }
            } else {
                result.addError("failed to parse tag.min as uint64: \(minStr)")
            }
        }
        
        if let maxStr = tag.max {
            if let maxi = Int(maxStr) {
                if maxi < 0 {
                    result.addError("tag.max \(maxi) is out of uint64 range [0, 18446744073709551615]")
                } else if intValue > maxi {
                    result.addError("value \(intValue) exceeds the maximum limit \(maxi)")
                }
            } else {
                result.addError("failed to parse tag.max as uint64: \(maxStr)")
            }
        }
    }

    private func validateBigInt(_ value: Any, tag: JSONCTag, result: ValidationResult) {
        guard let stringValue = value as? String else {
            result.addError("value must be a string")
            return
        }
        if stringValue.isEmpty && !tag.allowEmpty {
            result.addError("value is required")
            return
        }
        let bigIntRegex = "^-?[0-9]+$"
        let bigIntPredicate = NSPredicate(format: "SELF MATCHES %@", bigIntRegex)
        if !bigIntPredicate.evaluate(with: stringValue) {
            result.addError("value is not a valid bigInt")
        }
    }

    private func validateFloat32(_ value: Any, tag: JSONCTag, result: ValidationResult) {
        guard let floatValue = value as? Float else {
            result.addError("value must be a float")
            return
        }
        if floatValue == 0 && !tag.allowEmpty {
            result.addError("value is required")
            return
        }
        
        if let minStr = tag.min {
            if let mini = Float(minStr) {
                if mini < -3.402823466e+38 || mini > 3.402823466e+38 {
                    result.addError("tag.min \(mini) is out of float32 range [-3.402823466e+38, 3.402823466e+38]")
                } else if floatValue < mini {
                    result.addError("value \(floatValue) is less than the minimum limit \(mini)")
                }
            } else {
                result.addError("failed to parse tag.min as float32: \(minStr)")
            }
        }
        
        if let maxStr = tag.max {
            if let maxi = Float(maxStr) {
                if maxi < -3.402823466e+38 || maxi > 3.402823466e+38 {
                    result.addError("tag.max \(maxi) is out of float32 range [-3.402823466e+38, 3.402823466e+38]")
                } else if floatValue > maxi {
                    result.addError("value \(floatValue) exceeds the maximum limit \(maxi)")
                }
            } else {
                result.addError("failed to parse tag.max as float32: \(maxStr)")
            }
        }
    }

    private func validateFloat64(_ value: Any, tag: JSONCTag, result: ValidationResult) {
        guard let doubleValue = value as? Double else {
            result.addError("value must be a double")
            return
        }
        if doubleValue == 0 && !tag.allowEmpty {
            result.addError("value is required")
            return
        }
        
        if let minStr = tag.min {
            if let mini = Double(minStr) {
                if mini < -1.7976931348623157e+308 || mini > 1.7976931348623157e+308 {
                    result.addError("tag.min \(mini) is out of float64 range [-1.7976931348623157e+308, 1.7976931348623157e+308]")
                } else if doubleValue < mini {
                    result.addError("value \(doubleValue) is less than the minimum limit \(mini)")
                }
            } else {
                result.addError("failed to parse tag.min as float64: \(minStr)")
            }
        }
        
        if let maxStr = tag.max {
            if let maxi = Double(maxStr) {
                if maxi < -1.7976931348623157e+308 || maxi > 1.7976931348623157e+308 {
                    result.addError("tag.max \(maxi) is out of float64 range [-1.7976931348623157e+308, 1.7976931348623157e+308]")
                } else if doubleValue > maxi {
                    result.addError("value \(doubleValue) exceeds the maximum limit \(maxi)")
                }
            } else {
                result.addError("failed to parse tag.max as float64: \(maxStr)")
            }
        }
    }
    
    private func validateString(_ value: Any, tag: JSONCTag, result: ValidationResult) {
        if let strValue = value as? String {
            if strValue.isEmpty && !tag.allowEmpty {
                result.addError("value is empty")
            }
        } else {
            result.addError("value must be a string")
        }
    }
    
    private func validateEmail(_ value: Any, tag: JSONCTag, result: ValidationResult) {
        if let email = value as? String {
            if email.isEmpty && !tag.allowEmpty {
                result.addError("value is empty")
                return
            }
            
            if !email.isEmpty {
                let emailRegex = "^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\\.[a-zA-Z]{2,}$"
                let emailPredicate = NSPredicate(format: "SELF MATCHES %@", emailRegex)
                if !emailPredicate.evaluate(with: email) {
                    result.addError("value is not a valid email")
                }
            }
        } else {
            result.addError("value must be a string")
        }
    }
    
    private func validateUrl(_ value: Any, tag: JSONCTag, result: ValidationResult) {
        if let url = value as? String {
            if url.isEmpty && !tag.allowEmpty {
                result.addError("value is empty")
                return
            }
            
            if !url.isEmpty {
                if let urlObj = URL(string: url) {
                    if !urlObj.scheme!.starts(with: "http") {
                        result.addError("value is not a valid url")
                    }
                } else {
                    result.addError("value is not a valid url")
                }
            }
        } else {
            result.addError("value must be a string")
        }
    }
    
    private func validateBytes(_ value: Any, tag: JSONCTag, result: ValidationResult) {
        if let bytes = value as? Data {
            if bytes.isEmpty && !tag.allowEmpty {
                result.addError("value is empty")
            }
        } else {
            result.addError("value must be a byte array")
        }
    }
    
    private func validateUUID(_ value: Any, tag: JSONCTag, result: ValidationResult) {
        if let uuid = value as? String {
            if uuid.isEmpty && !tag.allowEmpty {
                result.addError("value is empty")
                return
            }
            
            if !uuid.isEmpty {
                if UUID(uuidString: uuid) == nil {
                    result.addError("value is not a valid uuid")
                }
            }
        } else {
            result.addError("value must be a string")
        }
    }
    
    private func validateDateTime(_ value: Any, tag: JSONCTag, result: ValidationResult) {
        if !(value is String) && !(value is Date) {
            result.addError("value must be a datetime")
        }
    }
    
    private func validateEnum(_ value: Any, tag: JSONCTag, result: ValidationResult) {
        if let enumValue = value as? String {
            if enumValue.isEmpty && !tag.allowEmpty {
                result.addError("value is empty")
                return
            }
            
            if !enumValue.isEmpty && !tag.enumValues.isEmpty {
                let enumValues = tag.enumValues.split(separator: "|").map { $0.trimmingCharacters(in: .whitespaces) }
                if !enumValues.contains(enumValue) {
                    result.addError("value is not in enum")
                }
            }
        } else {
            result.addError("value must be a string")
        }
    }
    
    private func validateArray(_ value: Any, tag: JSONCTag, result: ValidationResult) {
        if let array = value as? [Any] {
            if array.isEmpty && !tag.allowEmpty {
                result.addError("value is empty")
            }
            
            // 验证数组元素
            for (index, item) in array.enumerated() {
                let childTag = JSONCTag()
                childTag.inherit(from: tag)
                let itemResult = validate(item, tag: childTag)
                if !itemResult.isValid {
                    for error in itemResult.errors {
                        result.addError("item \(index): \(error)")
                    }
                }
            }
        } else {
            result.addError("value must be an array")
        }
    }
    
    private func validateStruct(_ value: Any, tag: JSONCTag, result: ValidationResult) {
        if !(value is [String: Any]) {
            result.addError("value must be an object")
        }
    }
}

public let validator = MmValidator()
